doc/design/service-discovery.md

```diff
@@ -0,0 +1,316 @@
+    * Notification-based callback is only called once and and user gets the complete service registry
+    * Polling with `FindService()` will always return the same result
+  * Alternative would be a `finalize()` method after init phase
+
+#### Sychronous, one-shot RPC (Polling)
+
+##### Alternative A: IPC channel usage
+
+Stick with IPC channel
+
+Pro:
+
+* No code changes needed
+* More memory efficient
+* No managment overhead on iceoryx-side
+
+Con:
+
+* No userspace solution
+  * IPC channel uses OS functions during runtime (switch between kernel and user space)
+* Upper frame limit may lead to segregation into several frames
+
+##### Alternative B: Request/ response feature usage
+
+* Re-use [request/ response feature](https://github.com/eclipse-iceoryx/iceoryx/issues/27)
+
+Pro:
+
+* Re-use building blocks to avoid code duplication
+* [Dogfooding](https://en.wikipedia.org/wiki/Eating_your_own_dog_food) your own features makes you aware of
+potential issues
+* Fast due to shared memory
+
+Con:
+
+* Shared memory approach might have overhead compared to a simple RPC call
+
+#### Event-based notification
+
+##### Alternative A: Shared memory + IPC channel
+
+* Send ID's over IPC channel and read ABA-counter/ServiceDiscoveryChangeCounter from shared memory
+* Data transfer over shared memory
+  * Basically putting Service Registry to shared memory so everyone can access it (read-only)
+
+Pro:
+
+* Fast data transfer
+* Could replace `InterfacePort`s
+
+Contra:
+
+* Bespoken solution more complex than todays approach (thread-safety required, lock-free mechanism needed)
+* In an optimal case IPC channel communication shall only be used for startup
+
+##### Alternative B: Built-in topic based on `InterfacePort`'s
+
+* Built-in topic approach based on `InterfacePort`'s
+  * Sniff and intercept `CaproMessage`
+* Also see
+  * [overview diagram](diagrams/service-discovery-sequence-diagram.puml)
+  * [sequence diagram](diagrams/service-discovery-overview.puml)
+
+Pro:
+
+* Close to DDS mechanism
+* `InterfacePort` already available
+* Synergies with gateway mechanisms
+* Gateways can benefit from event notification mechanism
+  * No need for a polling-loop anymore
+
+Con:
+
+* Overhead, `Listner` will wake up on every `CaproMessage`
+* Bookkeeping at user-side, just delta of service registry is transferred
+* What will RouDi do if he runs out of memory?
+  * Dimensioning according to max values is not optimal (MAX_INTERFACE_CAPRO_FIFO_SIZE)
+  * Presumably lots of memory needed, e.g. during startup phase when lots of apps will do discovery
+  * A safety-certified middleware based on iceoryx would not use the dynamic discovery feature
+
+Remark:
+
+* Need to filter out your own `CaproMessageType::FIND` requests
+* Extend `dispatchCaproMessage()` with condition variable and notification mechanism
+
+##### Alternative C: Custom thread
+
+* Implement a custom thread in `PoshRuntime`, which polls for specific discovery information using request/response
```

Comment:
This mainly makes sense if one does not want to use the listener and built-in topics. Otherwise the listener has this thread already and we just need to react on updates in its callback. I am not sure the user trigger callbacks are powerful to do so, but this can be corrected. They are able at least able to operate on some object or take an opaque void pointer, which is sufficient.

Reply:
Snapshoting is tricky in its own right. The registry container should not be protected by mutex, since contention would be quite high with a large number of applications (which will simultaneously try to get the snapshot if something has changed). Instead it must be able to be modified (insertions/deletion) by Roudi while those snapshots take place.

Not all apps will see the same snapshot, but this is ok as long as it is eventually consistent.

This is possible to do in a lock-free way, we may already have the right data structure, otherwise I could provide one I have already experimented with in another problem.

Reply:
My conclusion is that the effort to implement variant C is not reasonable and variant D is easier to implement. The potentially small performance does not justify not reusing the listener.